invoices/sql_migration_test.go

```diff
@@ -0,0 +1,434 @@
+
+	circutKey := models.CircuitKey{
+		ChanID: randChanID,
+		HtlcID: htlcID,
+	}
+
+	return circutKey, htlc
+}
+
+// generateInvoiceHTLCs generates all HTLCs including AMP HTLCs for an invoice.
+func generateInvoiceHTLCs(invoice *Invoice) {
+	mpp := invoice.Terms.Features.HasFeature(lnwire.MPPRequired)
+	numHTLCs := 1
+	if invoice.State == ContractOpen {
+		numHTLCs = 0
+	} else if mpp {
+		numHTLCs = rand.Intn(10)
+	}
+
+	total := invoice.Terms.Value
+
+	// Divide the total amount over the HTLCs equally.
+	if numHTLCs > 0 {
+		amt := total / lnwire.MilliSatoshi(numHTLCs)
+		remainder := total - amt*lnwire.MilliSatoshi(numHTLCs)
+		for i := 0; i < numHTLCs; i++ {
+			if i == numHTLCs-1 {
+				amt += remainder
+			}
+
+			circuitKey, htlc := randHTLC(invoice, amt)
+			invoice.Htlcs[circuitKey] = htlc
+		}
+	}
+}
+
+// generateAMPHtlcs generates AMP HTLCs for an invoice.
+func generateAMPHtlcs(t *testing.T, invoice *Invoice) {
+	// If this is an AMP invoice, we will add a few AMP HTLC sets as well.
+	numSetIDs := rand.Intn(5) + 1
+	settledIdx := uint64(1)
+
+	for i := 0; i < numSetIDs; i++ {
+		var setID SetID
+		_, err := crand.Read(setID[:])
+		require.NoError(t, err)
+
+		numHTLCs := rand.Intn(5) + 1
+		total := invoice.Terms.Value
+		invoiceKeys := make(map[CircuitKey]struct{})
+
+		var htlcState HtlcState
+		amt := total / lnwire.MilliSatoshi(numHTLCs)
+		remainder := total - amt*lnwire.MilliSatoshi(numHTLCs)
+
+		for j := 0; j < numHTLCs; j++ {
+			if j == numHTLCs-1 {
+				amt += remainder
+			}
+
+			// Generate the HTLC and the AMP data.
+			circuitKey, htlc := randHTLC(invoice, amt)
+			htlcState = htlc.State
+
+			var (
+				rootShare, hash [32]byte
+				preimage        lntypes.Preimage
+			)
+
+			_, err := crand.Read(rootShare[:])
+			require.NoError(t, err)
+
+			_, err = crand.Read(hash[:])
+			require.NoError(t, err)
+
+			_, err = crand.Read(preimage[:])
+			require.NoError(t, err)
+
+			record := record.NewAMP(rootShare, setID, uint32(j))
+
+			htlc.AMP = &InvoiceHtlcAMPData{
+				Record:   *record,
+				Hash:     hash,
+				Preimage: &preimage,
+			}
+
+			invoice.Htlcs[circuitKey] = htlc
+			invoiceKeys[circuitKey] = struct{}{}
+		}
+
+		ampState := InvoiceStateAMP{
+			State:       htlcState,
+			InvoiceKeys: invoiceKeys,
+		}
+		if htlcState == HtlcStateSettled {
+			// For simplicity we don't use a global unique settle
+			// index, but just increment it for each settled set.
+			ampState.SettleIndex = settledIdx
+			ampState.SettleDate = randInvoiceTime(invoice)
+			settledIdx++
+		}
+
+		// Given that we don't currently mix states we can assume that
+		// the invoice is fully paid if the AMP set is not canceled.
+		if htlcState != HtlcStateCanceled {
+			ampState.AmtPaid = invoice.Terms.Value
+		}
+
+		invoice.AMPState[setID] = ampState
+	}
+}
+
+// TestMigrateSingleInvoice tests the migration of a single invoice with random
+// data. Note that this test may need to be extended if the Invoice or any of
```

Comment:
Nit: We have always a newline and then 
// NOTE:

Reply:
Done.